File: 백준/Gold/5972. 택배 배송/택배 배송.java
import java.util.*;
import java.io.*;

public class Main {
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		st = new StringTokenizer(br.readLine());
		N = Integer.parseInt(st.nextToken());
		M = Integer.parseInt(st.nextToken());
		adj = new List[N+1];
		dist = new int[N+1];
		for(int i = 1; i<=N; i++) {
			adj[i] = new ArrayList<>();
			dist[i] = INF;
		}
		
		for(int i = 0; i <M;i++) {
			st = new StringTokenizer(br.readLine());
			int from = Integer.parseInt(st.nextToken());
			int to = Integer.parseInt(st.nextToken());
			int cost = Integer.parseInt(st.nextToken());
			adj[from].add(new Node(to, cost));
			adj[to].add(new Node(from, cost));
		}
		
		PriorityQueue<Node> pq = new PriorityQueue<>();
		dist[1] = 0;
		pq.add(new Node(1, 0));
		
		while(!pq.isEmpty()) {
			Node cur = pq.poll();
			
			if(dist[cur.idx] < cur.cost)
				continue;
			
			for(Node nxt : adj[cur.idx]) {
				if(dist[nxt.idx] > dist[cur.idx] + nxt.cost) {
					dist[nxt.idx] = dist[cur.idx] + nxt.cost;
					pq.add(new Node(nxt.idx, dist[nxt.idx]));
				}
			}
		}
		System.out.println(dist[N]);
	}
	static int N,M;
	static StringTokenizer st;
	static List<Node> adj[];
	static int dist[];
	static int INF = Integer.MAX_VALUE / 2;
	static class Node implements Comparable<Node>{
		int idx, cost;
		public Node(int idx, int cost) {
			this.idx = idx;
			this.cost = cost;
		}
		
		public int compareTo(Node o) {
			return cost - o.cost;
		}
	}
}

File: 백준/Gold/5972. 택배 배송/택배 배송.py
from heapq import heappush, heappop

class Node:
    def __init__(self, idx, cost):
        self.idx = idx
        self.cost = cost

    def __lt__(self, other):
        return self.cost < other.cost

    def __repr__(self):
        return f"({self.idx}, {self.cost})"

if __name__ == '__main__':
    N,M = map(int,input().split())
    INF = float('inf')
    adj = [[] for _ in range(N+1)]
    dist = [INF for _ in range(N+1)]

    for _ in range(M):
        u,v,cost = map(int,input().split())
        adj[u].append(Node(v,cost))
        adj[v].append(Node(u,cost))

    pq = []
    dist[1] = 0
    heappush(pq, Node(1, 0))

    while pq:
        cur = heappop(pq)

        if dist[cur.idx] < cur.cost:
            continue

        for nxt in adj[cur.idx]:
            if dist[nxt.idx] > dist[cur.idx] + nxt.cost:
                dist[nxt.idx] = dist[cur.idx] + nxt.cost
                heappush(pq, Node(nxt.idx, dist[nxt.idx]))

    print(dist[N])